### PR TITLE
fix: include gate and measurement noise

### PR DESCRIPTION
**Pull request by notmgsk**
This MR
- adds the required gate and measurement noise parameters
- updates the README

### CODE REVIEW

File: examples/qvm/multishot.c [hunk 1]
```diff
@@ -82,7 +82,8 @@ void multishot_with_implicit_ro_indices() {
 
   qvm_multishot_result qvm_res;
   int num_trials = 10;
-  if (qvm_multishot(program, addresses, num_trials, &qvm_res) !=
+  double gate_noise[] = {0.0, 0.0, 0.0};
+  if (qvm_multishot(program, addresses, num_trials, &qvm_res, NULL, NULL) !=
```
Comment (kalzoo) [inline examples/qvm/multishot.c:86]: Curious why the arg ordering is different than L39 above?

And neither appears to match the signature defined in the Lisp change in this PR but I assume there's an FFI reason for that

File: examples/qvm/multishot.c [hunk 2]
```diff
@@ -108,11 +109,60 @@ void multishot_with_implicit_ro_indices() {
...
+  lisp_release_handle(qvm_res);
+  lisp_release_handle(program);
+}
+
 int main(int argc, char **argv) {
   init("../../libquil.core");
 
   multishot_with_explicit_ro_indices();
-  multishot_with_implicit_ro_indices();
+  // multishot_with_implicit_ro_indices();
```
Comment (kalzoo) [inline examples/qvm/multishot.c:165]: Did this start failing? I don't see any changes to this function here

File: examples/qvm/multishot.c [hunk 3]
```diff
@@ -108,11 +109,60 @@ void multishot_with_implicit_ro_indices() {
...
+  }
+
+  lisp_release_handle(qvm_res);
+  lisp_release_handle(program);
+}
+
 int main(int argc, char **argv) {
   init("../../libquil.core");
 
   multishot_with_explicit_ro_indices();
```
Comment (kalzoo) [inline examples/qvm/multishot.c:164]: Should `multishot_with_noise` be called here?

**Review by kalzoo — COMMENTED** [review]
The [issue](https://github.com/rigetti/libquil/issues/44) has an empty description and so does this MR so I'm not sure how to approach it, but it appears that there were a couple other args that can be threaded through the FFI layer and that does that here, with a single test?

Is there a pattern we could use to avoid breaking changes when arguments are changed - like a builder pattern? so `qvm_multishot` remains stable over time without breaking changes like this PR

**Comment by notmgsk** [comment]
> The [issue](https://github.com/rigetti/libquil/issues/44) has an empty description and so does this MR so I'm not sure how to approach it, but it appears that there were a couple other args that can be threaded through the FFI layer and that does that here, with a single test?

I intended to open this as a draft MR. Not sure why sometimes draft is the default and other times it is not.

File: examples/qvm/multishot.c (hunk 1 above)
Comment (notmgsk) [reply examples/qvm/multishot.c]: That particular example was incorrect (the result always comes in the last position). In the Lisp code, the result is not always included as a parameter in the function signature, in which case there is an implicit argument which takes its value from the result of the corresponding Lisp function.

**Comment by notmgsk** [comment]
> Is there a pattern we could use to avoid breaking changes when arguments are changed - like a builder pattern? so qvm_multishot remains stable over time without breaking changes like this PR

I had discussed this with folks a while back, e.g. changing `multishot(program, addresses, num_trials, NULL, NULL, &result)` to `multishot(request, &result)` but had difficulty making that work in an ergonomic way. (If everything was in C, it would've been fairly simple, but handing that to (or back from) Lisp was more complicated). It's also typical, if unsightly, to see C functions like this with many arguments. From a BC perspective, I think this is the only incompatibility.

**Pull request merged**